instance/github.py

```diff
@@ -0,0 +1,164 @@
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+GitHub Service API - Helper functions
+"""
+
+# Imports #####################################################################
+
+import re
+import requests
+
+from django.conf import settings
+
+
+# Logging #####################################################################
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+# Constants ###################################################################
+
+GH_HEADERS = {
+    'Authorization': 'token {}'.format(settings.GITHUB_ACCESS_TOKEN),
+    'Time-Zone': 'UTC',
+}
+
+
+# Functions ###################################################################
+
+def get_object_from_url(url):
+    """
+    Send the request to the provided URL, attaching custom headers, and returns
+    the deserialized object from the returned JSON
+    """
+    logger.info('GET URL %s', url)
+    r = requests.get(url, headers=GH_HEADERS)
+    r.raise_for_status()
+    return r.json()
+
+
+def fork_name2tuple(fork_name):
+    """
+    Converts a `fork_name` (eg. `'open-craft/edx-platform'`)
+    to a `fork_tuple` (eg. `['open-craft', 'edx-platform']`)
+    """
+    return fork_name.split('/')
+
+
+def get_commit_id_from_ref(fork_name, ref_name, ref_type='heads'):
+    """
+    Get the `commit_id` currently attached to a git reference
+    """
+    url = 'https://api.github.com/repos/{fork_name}/git/refs/{ref_type}/{ref_name}'.format(
+        fork_name=fork_name,
+        ref_type=ref_type,
+        ref_name=ref_name,
+    )
+    return get_object_from_url(url)['object']['sha']
+
+
+def get_settings_from_pr_body(pr_body):
+    """
+    Extract a settings string from a PR description body
+    """
+    m = re.search("..Settings..\r?\n+```[a-z]*\r?\n((?:.+\r?\n)*)```", pr_body)
+    if m:
+        return m.groups()[0]
+    else:
+        return ''
+
+
+def get_pr_by_number(fork_name, pr_number):
+    """
+    Returns a PR() namedtuple based on the reponse
+    """
+    r_pr = get_object_from_url('https://api.github.com/repos/{fork_name}/pulls/{pr_number}'.format(
+        fork_name=fork_name,
+        pr_number=pr_number,
+    ))
+    pr_fork_name = r_pr['head']['repo']['full_name']
+    pr_branch_name = r_pr['head']['ref']
+    pr = PR(
+        pr_number,
+        pr_fork_name,
+        pr_branch_name,
+        r_pr['title'],
+        r_pr['user']['login'],
+        body=r_pr['body'],
+    )
+    return pr
+
+
+def get_pr_list_from_username(user_name, fork_name):
+    """
+    Retreive the current active PRs for a given user
```

Comment:
@bradenmacdonald Thanks - that's one of these words my brain registered with the wrong spelling : ) Fixed throughout the code base.